feat: add support for auto-complete of annotations

paulbutcher opened the pull request.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/LanguageServer.scala:
```diff
@@ -381,7 +381,7 @@ class LanguageServer(port: Int) extends WebSocketServer(new InetSocketAddress("l
     */
   private def wordAt(s: String, n: Int): Option[String] = {
-    def isValidChar(c: Char): Boolean = Character.isLetterOrDigit(c) || c == '.' || c == '/'
+    def isValidChar(c: Char): Boolean = Character.isLetterOrDigit(c) || c == '.' || c == '/' || c == '@'
```
This seems reasonable.

magnus-madsen commented:
You had some longer description on Gitter or was it here on GitHub? Anyway, could you post a quick description of the change?

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/CompleteProvider.scala:
```diff
@@ -33,13 +33,41 @@ object CompleteProvider {
-      getKeywordCompletionItems(line, word) ++
-      getSnippetCompletionItems(line, word)
+    word match {
+      case None => Nil;
```
What's the idea here? I guess it is not to offer suggestions if there is no text (and vscode asks for completions).

But on the other hand, if there is no text and the user presses ctrl+space then I guess nothing happens now?

magnus-madsen replied:
So I am not exactly sure.

paulbutcher replied:
It's more that previously the code didn't look at `word` at all, but now it does. Given that it does now look at `word` it has to do *something* when it's not defined...

magnus-madsen replied:
OK; I propose to structure it like this:

```
if (isAnnotationPrefix(word) { 
   ... getAnnotationCompletionItems() ...
} else {
  old block
}
```

and then add a helper function

```
def isAnnotationPrefix(word: Option[String]): Boolean
```

and go from there. 

Makes sense?

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/CompleteProvider.scala:
```diff
@@ -33,13 +33,41 @@ object CompleteProvider {
+    word match {
+      case None => Nil;
+      case Some(w) => 
+        if (w.startsWith("@")) {
```
This seems reasonable, but I would move the startsWith check into getAnnotationCompletionItems

magnus-madsen replied:
Ah- on second thought, Its because you don't want to give the other suggestions when writing `@...` Hmm.

magnus-madsen replied:
Perhaps we need a helper function? `isAnnotationPrefix` or whatever. Not sure.

paulbutcher replied:
Yeah, you understand my thinking here. Happy to structure it however you see fit.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/CompleteProvider.scala:
```diff
@@ -33,13 +33,41 @@ object CompleteProvider {
+      CompletionItem("Space", "Space ", None, Some("annotation"), CompletionItemKind.Keyword, InsertTextFormat.PlainText, Nil),
+      CompletionItem("Time", "Time ", None, Some("annotation"), CompletionItemKind.Keyword, InsertTextFormat.PlainText, Nil),
+      CompletionItem("Unsafe", "Unsafe ", None, Some("annotation"), CompletionItemKind.Keyword, InsertTextFormat.PlainText, Nil)
+    )
```
The above approach is reasonable, but I would remove the following:

- Internal
- Unsafe

since we are not sure where we are going with those.

magnus-madsen replied:
Oh, and I tihnk all of the completion items should start with `@`

paulbutcher replied:
Regarding starting with `@`, I agree that that's probably how it should be, but as I said in Gitter, if I do that then when the completion is inserted by VSCode, the @ ends up being doubled.

My guess is that something within SemanticTokensProvider probably needs updating (I think that that's where VSCode is getting its tokenisation from?). But I'm not sure where to start here?

magnus-madsen replied:
I am not sure what the solution is, but the SemanticTokensProvider is completely separate from auto complete.

magnus-madsen replied:
(i.e. they have no interaction whatsoever in VSCode)

paulbutcher replied:
So this code simply returns a list of `CompletionItem`s (it ignores the parameters). So presumably VSCode is doing some kind of tokenisation and matching to decide which of the `CompletionItem`s are applicable?

Right now, it doesn't seem to be considering an `@` character to be part of the token under consideration? Presumably something is telling VSCode what constitutes a token for Flix? I can't find anything within https://github.com/flix/vscode-flix which defines this, so either it's something within VSCode itself, or something within the language server?

magnus-madsen replied:
We have often found that VSCode makes all kinds of "interpretations". You may think its just an API, but often it will "re-interpret" the data you are sending back. Its possible that it treats `@` differently in some kind of implementation specific way. (its also possible we have a mistake). But this would not be first time that VSCode does something slightly different than from what you would think the spec says.

magnus-madsen replied:
(Auto-completion does not use any kind of tokenization really).

magnus-madsen replied:
What happens if you change "Some(annotation)" to "Some(keyword)" does it behave differently?

paulbutcher replied:
Yeah, I tried changing `Some(annotation)`. It makes no functional difference (it's just help text that's displayed if you click on the `>` in the UI).

magnus-madsen replied:
Ok

paulbutcher commented:
Sorry - I created this in a rush - I should have waited until I had more time so I could include a proper description.

I've responded to your comments above. The bottom line is that I think that there need to be some accompanying changes to `SemanticTokensProvider` but I don't really know where to start here. Guidance welcome!

(there's a reason why I made this a draft - I'm not entirely happy with it).

paulbutcher commented:
I've checked in code which I believe addresses your comments above (i.e. removing `Internal` and `Unsafe` and restructuring). Let me know if I've missed anything.

magnus-madsen commented:
Thanks!
